src/lib.rs

```diff
@@ -363,253 +527,465 @@ struct Category<BlockNumber, Moment, AccountId> {
-        // without worrying about gameability or attack scenarios.
-        // If you not specify `Result` explicitly as return value, it will be added automatically
-        // for you and `Ok(())` will be returned.
-        fn set_dummy(#[compact] new_value: BalanceOf<T>) {
-            // Put the new value into storage.
-            <Dummy<T>>::put(new_value);
+        /// Add a new category.
+        fn create_category(origin, parent: Option<CategoryId>, title: Vec<u8>, description: Vec<u8>) -> dispatch::Result {
+
+            // Check that its a valid signature
+            let who = ensure_signed(origin)?;
+
+            // Not signed by forum SUDO
+            Self::ensure_is_forum_sudo(&who)?;
+
+            // Validate title
+            ensure_category_title_is_valid(&title)?;
+
+            // Validate description
+            ensure_category_description_is_valid(&description)?;
+
+            // If not root, then check that we can create in parent category
+            if let Some(parent_category_id) = parent {
+
+                // Get path from parent to root of category tree.
+                let mut category_tree_path = vec![];
+
+                Self::build_category_tree_path(parent_category_id, &mut category_tree_path);
+
+                // Can we mutate in this category?
+                Self::ensure_can_add_subcategory_to_parent_at_path_tip(&category_tree_path)?;
+            }
+
+            let next_category_id = <NextCategoryId<T>>::get();
+
+            // Create new category
+            let new_category = Category {
+                id : next_category_id,
+                title : title.clone(),
+                description: description.clone(),
+                created_at : Self::current_block_and_time(),
+                deleted: false,
+                archived: false,
+                num_threads_ever_created: 0,
+                parent_id: parent,
+                moderator_id: who
+            };
+
+            // Insert category in map
+            <CategoryById<T>>::insert(new_category.id, new_category);
+
+            // Update other things
+            <NextCategoryId<T>>::put(next_category_id + 1);
+
+            // Generate event
+            Self::deposit_event(RawEvent::CategoryCreated(next_category_id));
+
+            Ok(())
         }
+        
+        /// Update category
+        fn update_category(origin, category_id: CategoryId, archived: bool, deleted: bool) -> dispatch::Result {
+
+            // Check that its a valid signature
+            let who = ensure_signed(origin)?;
+
+            // Not signed by forum SUDO
+            Self::ensure_is_forum_sudo(&who)?;
+
+            // Get path from parent to root of category tree.
+            let mut category_tree_path = vec![];
+
+            Self::build_category_tree_path(category_id, &mut category_tree_path);
+
+            if !category_tree_path.is_empty() {
```

Comment:
you can check that category exists, and if not, the return from this function.
It will allow you to move the current block of `if` statement out of `if`.

Pseudocode:

```rust
if category_does_not_exist {
    // If path is empty, it just means the category did not exist,
    // which means the user provided an invalid category id.
    Err(ERROR_CATEGORY_DOES_NOT_EXIST)
}

// update the found category here:
// ....
```

Reply:
Discarded, premature review.